src/encryption.rs

```diff
@@ -0,0 +1,167 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at http://mozilla.org/MPL/2.0/.
+//
+// Copyright (c) DUSK NETWORK. All rights reserved.
+
+use alloc::vec::Vec;
+
+use crate::{Call, Error, Safe, Sponge};
+use zeroize::Zeroize;
+
+/// Trait that together with the [`Safe`] trait, implements everything that is
+/// needed for encryption using the SAFE framework.
+///
+/// The specific implementation of subtraction and equality is needed for being
+/// able to use the trait in a zero-knowledge circuit.
+pub trait Encryption<T, const W: usize> {
+    /// Specifically implement subtraction for the type `T`:
+    /// minuend - subtrahend = difference
+    fn subtract(&mut self, minuend: &T, subtrahend: &T) -> T;
```

Comment:
We might want to use the `Sub` trait for this instead of essentially redefining it here.

Reply:
P.S. did not know that english used "minuend" and "subtrahend"

Reply:
We can't use `Sub` because it won't work in a plonk circuit with `Witnesses`.

Reply:
this is also why there is a mut ref to self: this will give us the composer